bolt-modules/boltlib/lib/puppet/functions/wait.rb

```diff
@@ -58,22 +87,30 @@
   end
 
   def wait(futures, options = {})
-    inner_wait(futures, nil, options)
+    inner_wait(futures: futures, options: options)
+  end
+
+  def wait_for_all(options = {})
+    inner_wait(options: options)
+  end
+
+  def wait_for_all_with_timeout(timeout, options = {})
+    inner_wait(timeout: timeout, options: options)
   end
 
   def wait_with_timeout(futures, timeout, options = {})
-    inner_wait(futures, timeout, options)
+    inner_wait(futures: futures, timeout: timeout, options: options)
   end
 
-  def inner_wait(futures, timeout = nil, options = {})
+  def inner_wait(futures: nil, timeout: nil, options: {})
```

Comment:
What's the difference between using `=` or `:`  here? It seems that they both set default values for the arguments.

Reply:
Yeah `:` makes these [keyword arguments](https://thoughtbot.com/upcase/videos/ruby-keyword-arguments) or "named parameters" instead of just optional parameters. The main difference is when you call the functions.

With optional parameters (`=`), the parameter is optional (has a default value) but when you call the function you still have to know what parameters the function takes and position the arguments correctly to call the function. This makes calling the function confusing if you don't already know the function definition.
```
def function(breakfast = 'oatmeal', lunch = 'a sandwich')
  puts "Today I had #{breakfast} for breakfast and #{lunch} for lunch"
end

function() # Uses the defaults
function('waffles') # Sets 'waffles' as breakfast
function('oatmeal', 'tacos') # We have to manually pass in the first default so that Ruby knows we're setting *lunch*, not breakfast. Plus, you have to infer or Just Know that you had oatmeal for breakfast and tacos for lunch, and not the other way around.
```

Whereas with named arguments the function calls are self documenting so we know what parameters are being passed, and we can pass them in any order (and even do fancy things like [make a hash of arguments and pass them all at once](https://www.justinweiss.com/articles/fun-with-keyword-arguments/).
```
def function(breakfast: 'oatmeal', lunch: 'a sandwich')
  puts "Today I had #{breakfast} for breakfast and #{lunch} for lunch"
end

function() # Uses the defaults
function(breakfast: 'waffles') # Yum
function(lunch: 'tacos') # Such readability, so wow
```

Honestly, this wasn't something I really thought about until recently so I've just been updating our methods as I make changes. We can't just do it willy-nilly since some other tools use our functions directly or override them, but here it's safe to rewrite the definition.